sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/UnsafeArrayDataBenchmark.scala

```diff
@@ -0,0 +1,298 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.benchmark
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.catalyst.expressions.{UnsafeArrayData, UnsafeRow}
+import org.apache.spark.sql.catalyst.expressions.codegen.{BufferHolder, UnsafeArrayWriter}
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.util.Benchmark
+
+/**
+ * Benchmark [[UnsafeArrayDataBenchmark]] for UnsafeArrayData
+ * To run this:
+ *  build/sbt "sql/test-only *benchmark.UnsafeArrayDataBenchmark"
+ *
+ * Benchmarks in this file are skipped in normal builds.
+ */
+class UnsafeArrayDataBenchmark extends BenchmarkBase {
+
+  new SparkConf()
+    .setMaster("local[1]")
+    .setAppName("microbenchmark")
+    .set("spark.driver.memory", "3g")
+
+  def calculateHeaderPortionInBytes(count: Int) : Int = {
+    // Use this assignment for SPARK-15962
+    // val size = 4 + 4 * count
+    val size = UnsafeArrayData.calculateHeaderPortionInBytes(count)
+    size
+  }
+
+  def readUnsafeArray(iters: Int): Unit = {
```

Comment:
I think we need to benchmark 4 cases:

1. normal write: generate a random int array and use encoder to turn it into array data, e.g.
```
val array: Array[Int] = ...
val encoder = ExpressionEncoder[Array[Int]].resolveAndBind()
encoder.toRow(array) // benchmark it.
```
2. from primitive array: benchmark the `UnsafeArrayData.fromPrimitiveArray`
3. normal read: generate random array, turn it into array data by encoder, then benchmark the element reading, e.g. `getInt`
4. to primitive array: benchmark the `UnsafeArrayData.toIntArray`, etc.